Create vm report: * vsphere_virtual_machine.vm: could not find resource pool ID ""

_This issue was originally opened by @liucj0120 as hashicorp/terraform#18102. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,when i create a vmware virtual machine,report a error about "resource pool ID"
# terraform version
Terraform v0.11.7
+ provider.vsphere v1.5.0

Error information:
```
data.vsphere_datacenter.dc: Refreshing state...
data.vsphere_host.hosts: Refreshing state...
data.vsphere_datastore.datastore: Refreshing state...
data.vsphere_virtual_machine.template: Refreshing state...

Error: Error running plan: 1 error(s) occurred:

* vsphere_virtual_machine.vm: 1 error(s) occurred:

* vsphere_virtual_machine.vm: could not find resource pool ID "": ServerFaultCode: The object 'vim.ResourcePool:' has already been deleted or has not been completely created
```

Terraform  Configuration Files
vm.tf:

```
provider "vsphere" {
  user           = "${var.user}"
  password       = "${var.password}"
  vsphere_server = "${var.vsphere_server}"

 # If you have a self-signed cert
  allow_unverified_ssl = true
}

#variable "datacenter" {
#  default = "Datacenter"
#}

#variable "hosts" {
#  default = [
#    "172.29.254.200",
#  ]
#}

#variable "vswitch" {
#  default = "vSwitchterraform"
#}

#variable "portgroup" {
#  default = "PGterraform"
#}

#variable "vmname" {
#  default = "terraform-test"
#}

data "vsphere_datacenter" "dc" {
  name = "${var.datacenter}"
}

data "vsphere_host" "hosts" {
  count         = "${length(var.hosts)}"
  name          = "${var.hosts[count.index]}"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_compute_cluster" "cluster" {
  name            = "terraform-cluster"
  datacenter_id   = "${data.vsphere_datacenter.dc.id}"
  host_system_ids = ["${data.vsphere_host.hosts.*.id}"]

  drs_enabled          = true
  drs_automation_level = "fullyAutomated"

  ha_enabled = true
}

resource "vsphere_host_virtual_switch" "switch" {
  name           = "${var.vswitch}"
  host_system_id = "${data.vsphere_host.hosts.id}"

  network_adapters = ["vmnic1", "vmnic2"]

  active_nics  = ["vmnic1"]
  standby_nics = ["vmnic2"]
}

resource "vsphere_host_port_group" "pg" {
  name                = "${var.portgroup}"
  host_system_id      = "${data.vsphere_host.hosts.id}"
  virtual_switch_name = "${vsphere_host_virtual_switch.switch.name}"

  vlan_id = 100

  allow_promiscuous = true
}

#create vm

#data "vsphere_compute_cluster" "cluster" {
#  name          = "terraform-cluster"
#  datacenter_id = "${data.vsphere_datacenter.dc.id}"
#}

data "vsphere_datastore" "datastore" {
  name          = "datastore1"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_virtual_machine" "template" {
  name          = "centos7template"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_virtual_machine" "vm" {
  name             = "${var.vmname}"
  resource_pool_id = "${vsphere_compute_cluster.cluster.resource_pool_id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"

  num_cpus = 4
  memory   = 2048
  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  scsi_type = "${data.vsphere_virtual_machine.template.scsi_type}"

  network_interface {
    network_id   = "${vsphere_host_port_group.pg.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
    label            = "disk0"
    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

    customize {
      linux_options {
        host_name = "${var.vmname}"
        domain    = "test.internal"
      }

      network_interface {
        ipv4_address = "172.29.254.135"
        ipv4_netmask = 24
      }

      ipv4_gateway = "172.29.254.1"
    }
  }
}
```



Variable configuration file:
variable.tf:

```
variable "user" {
  default = "administrator@vsphere.local"
}

variable "password" {
#  default = "A@1234cn"
}

variable "vsphere_server" {
  default = "172.29.254.133"
}

variable "datacenter" {
  default = "Datacenter"
}

variable "hosts" {
  default = [
    "172.29.254.200",
  ]
}

variable "vswitch" {
  default = "vSwitchterraform"
}

variable "portgroup" {
  default = "PGterraform"
}

variable "vmname" {
  default = "terraform-test"
}

```

Thanks.

I'm fighting with same problem right now.
After examining debug logs, I've realized that `resourceVSphereVirtualMachineCustomizeDiff` is being called 4 times for the one vm instance I have.
When it's called first time, the value of `resource_pool_id` is not yet `computed`, so it fails without being able to find it:
```
2018/06/11 14:49:49 [ERROR] root.win-iacdev-prn: eval: *terraform.EvalSequence, err: could not find resource pool ID "": ServerFaultCode: The object 'vim.ResourcePool:' has already been deleted or has not been completely created
```
This error is not consistent and depends on how Terraform builds and executes graph.

--- a/vsphere/internal/vmworkflow/virtual_machine_clone_subresource.go
+++ b/vsphere/internal/vmworkflow/virtual_machine_clone_subresource.go
@@ -97,17 +97,20 @@ func ValidateVirtualMachineClone(d *schema.ResourceDiff, c *govmomi.Client) erro
 
 	// If a customization spec was defined, we need to check some items in it as well.
 	if len(d.Get("clone.0.customize").([]interface{})) > 0 {
-		poolID := d.Get("resource_pool_id").(string)
-		pool, err := resourcepool.FromID(c, poolID)
-		if err != nil {
-			return fmt.Errorf("could not find resource pool ID %q: %s", poolID, err)
-		}
-		family, err := resourcepool.OSFamily(c, pool, d.Get("guest_id").(string))
-		if err != nil {
-			return fmt.Errorf("cannot find OS family for guest ID %q: %s", d.Get("guest_id").(string), err)
-		}
-		if err := ValidateCustomizationSpec(d, family); err != nil {
-			return err
+		if poolID, ok := d.GetOk("resource_pool_id"); ok {
+			pool, err := resourcepool.FromID(c, poolID.(string))
+			if err != nil {
+				return fmt.Errorf("could not find resource pool ID %q: %s", poolID, err)
+			}
+			family, err := resourcepool.OSFamily(c, pool, d.Get("guest_id").(string))
+			if err != nil {
+				return fmt.Errorf("cannot find OS family for guest ID %q: %s", d.Get("guest_id").(string), err)
+			}
+			if err := ValidateCustomizationSpec(d, family); err != nil {
+				return err
+			}
+		} else {
+			log.Printf("[DEBUG] ValidateVirtualMachineClone: resource_pool_id is not available. Skipping OS family check.")
 		}
 	}
 	vconfig := vprops.Config.VAppConfig

--- a/vsphere/resource_vsphere_virtual_machine_test.go
+++ b/vsphere/resource_vsphere_virtual_machine_test.go
@@ -1446,6 +1446,25 @@ func TestAccResourceVSphereVirtualMachine_cloneNoGateway(t *testing.T) {
 	})
 }
 
+func TestAccResourceVSphereVirtualMachine_cloneCustomizeWithNewResourcePool(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccResourceVSphereVirtualMachinePreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccResourceVSphereVirtualMachineCheckExists(false),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccResourceVSphereVirtualMachineConfigCloneWithNewResourcePool(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccResourceVSphereVirtualMachineCheckExists(true),
+				),
+			},
+		},
+	})
+}
+
 func TestAccResourceVSphereVirtualMachine_cloneCustomizeForceNew(t *testing.T) {
 	var state *terraform.State
 
@@ -7009,6 +7028,135 @@ resource "vsphere_virtual_machine" "vm" {
 		os.Getenv("VSPHERE_RESOURCE_POOL"),
 		os.Getenv("VSPHERE_NETWORK_LABEL_PXE"),
 		os.Getenv("VSPHERE_DATASTORE"),
+	)
+}
+
+func testAccResourceVSphereVirtualMachineConfigCloneWithNewResourcePool() string {
+	return fmt.Sprintf(`
+variable "datacenter" {
+  default = "%s"
+}
+
+variable "resource_pool" {
+  default = "%s"
+}
+
+variable "network_label" {
+  default = "%s"
+}
+
+variable "ipv4_address" {
+  default = "%s"
+}
+
+variable "ipv4_netmask" {
+  default = "%s"
+}
+
+variable "ipv4_gateway" {
+  default = "%s"
+}
+
+variable "dns_server" {
+  default = "%s"
+}
+
+variable "datastore" {
+  default = "%s"
+}
+
+variable "template" {
+  default = "%s"
+}
+
+variable "linked_clone" {
+  default = "%s"
+}
+
+data "vsphere_datacenter" "dc" {
+  name = "${var.datacenter}"
+}
+
+data "vsphere_datastore" "datastore" {
+  name          = "${var.datastore}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_resource_pool" "pool" {
+  name          = "${var.resource_pool}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_network" "network" {
+  name          = "${var.network_label}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_virtual_machine" "template" {
+  name          = "${var.template}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+resource "vsphere_resource_pool" "pool" {
+  name                    = "terraform-test-resource-pool"
+  parent_resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
+}
+
+resource "vsphere_virtual_machine" "vm" {
+  name             = "terraform-test"
+  resource_pool_id = "${vsphere_resource_pool.pool.id}"
+  datastore_id     = "${data.vsphere_datastore.datastore.id}"
+
+  num_cpus = 2
+  memory   = 2048
+  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"
+
+  wait_for_guest_net_timeout = -1
+
+  network_interface {
+    network_id   = "${data.vsphere_network.network.id}"
+    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
+  }
+
+  disk {
+    label            = "disk0"
+    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
+    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
+    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
+  }
+
+  clone {
+    template_uuid = "${data.vsphere_virtual_machine.template.id}"
+    linked_clone  = "${var.linked_clone != "" ? "true" : "false" }"
+
+    customize {
+      linux_options {
+        host_name = "terraform-test"
+        domain    = "test.internal"
+      }
+
+      network_interface {
+        ipv4_address = "${var.ipv4_address}"
+        ipv4_netmask = "${var.ipv4_netmask}"
+      }
+
+      ipv4_gateway    = "${var.ipv4_gateway}"
+      dns_server_list = ["${var.dns_server}"]
+      dns_suffix_list = ["test.internal"]
+    }
+  }
+}
+`,
+		os.Getenv("VSPHERE_DATACENTER"),
+		os.Getenv("VSPHERE_RESOURCE_POOL"),
+		os.Getenv("VSPHERE_NETWORK_LABEL"),
+		os.Getenv("VSPHERE_IPV4_ADDRESS"),
+		os.Getenv("VSPHERE_IPV4_PREFIX"),
+		os.Getenv("VSPHERE_IPV4_GATEWAY"),
+		os.Getenv("VSPHERE_DNS"),
+		os.Getenv("VSPHERE_DATASTORE"),
+		os.Getenv("VSPHERE_TEMPLATE"),
+		os.Getenv("VSPHERE_USE_LINKED_CLONE"),
 	)
 }
 

